tests/Order.Service.Tests/MicrocksWebApplicationFactory.cs

```diff
@@ -67,57 +86,125 @@ private ushort GetAvailablePort()
-            .WithAsyncFeature() // We need this for async mocking and contract-testing
-            .WithPostman() // We need this for Postman contract-testing
-            .WithMainArtifacts("resources/order-service-openapi.yaml", "resources/order-events-asyncapi.yaml", "resources/third-parties/apipastries-openapi.yaml")
-            .WithSecondaryArtifacts("resources/order-service-postman-collection.json", "resources/third-parties/apipastries-postman-collection.json")
-            .WithKafkaConnection(new KafkaConnection(kafkaListener)); // We need this to connect to Kafka
-
-        await this.MicrocksContainerEnsemble.StartAsync()
-            .ConfigureAwait(true);
+        // Use semaphore to ensure only one initialization happens across all test instances
+        await InitializationSemaphore.WaitAsync(TestContext.Current.CancellationToken);
+        try
+        {
+            if (_isInitialized)
+            {
+                TestLogger.WriteLine("[MicrocksWebApplicationFactory] Factory already initialized, skipping...");
+                return;
+            }
+
+            TestLogger.WriteLine("[MicrocksWebApplicationFactory] Starting initialization...");
+
+            // The port is dynamically determined because we use Microcks,
+            // so we need to get an available port before starting the server (Kestrel) and Microcks.
+            // because we use microcks to set up the base address for the API in the settings.
+            ActualPort = GetAvailablePort();
+            TestLogger.WriteLine("[MicrocksWebApplicationFactory] Using port: {0}", ActualPort);
+
+            UseKestrel(ActualPort);
+            await TestcontainersSettings.ExposeHostPortsAsync(ActualPort, TestContext.Current.CancellationToken)
+                .ConfigureAwait(true);
+
+            string kafkaListener = "kafka:19092";
+
+            var network = new NetworkBuilder().Build();
+            TestLogger.WriteLine("[MicrocksWebApplicationFactory] Creating Kafka container...");
+
+            KafkaContainer = new KafkaBuilder()
+                .WithImage("confluentinc/cp-kafka:7.9.0")
+                .WithPortBinding(9092, KafkaBuilder.KafkaPort)
+                .WithPortBinding(9093, KafkaBuilder.BrokerPort)
+                .WithNetwork(network)
+                .WithNetworkAliases("kafka")
+                .WithListener(kafkaListener)
+                .Build();
+
+            // Start the Kafka container
+            TestLogger.WriteLine("[MicrocksWebApplicationFactory] Starting Kafka container...");
+            await this.KafkaContainer.StartAsync(TestContext.Current.CancellationToken)
+                .ConfigureAwait(true);
+
+            // Create the Microcks container ensemble with the Kafka connection
+            this.MicrocksContainerEnsemble = new MicrocksContainerEnsemble(network, MicrocksImage)
+                .WithAsyncFeature() // We need this for async mocking and contract-testing
```

Comment:
You could also specify the `quay.io/microcks/microcks-uber-async-minion:1.13.0-native` for an even faster bootstrap.